localstack-core/localstack/utils/backoff.py

```diff
@@ -0,0 +1,97 @@
+import random
+import time
+
+from pydantic import Field
+from pydantic.dataclasses import dataclass
+
+
+@dataclass
+class ExponentialBackoff:
+    """
+    ExponentialBackoff implements exponential backoff with randomization.
+    The backoff period increases exponentially for each retry attempt, with
+    optional randomization within a defined range.
+
+    next_backoff() is calculated using the following formula:
+        ```
+        randomized_interval = random_between(retry_interval * (1 - randomization_factor), retry_interval * (1 + randomization_factor))
+        ```
+
+    For example, given:
+        `initial_interval` = 2
+        `randomization_factor` = 0.5
+        `multiplier` = 2
+
+    The next backoff will be between 1 and 3 seconds (2 * [0.5, 1.5]).
+    The following backoff will be between 2 and 6 seconds (4 * [0.5, 1.5]).
+
+    Note:
+        - `max_interval` caps the base interval, not the randomized value
+        - Returns 0 when `max_retries` or `max_time_elapsed` is exceeded
+        - The implementation is not thread-safe
+
+    Example sequence with defaults (initial_interval=0.5, randomization_factor=0.5, multiplier=1.5):
+
+    | Request # | Retry Interval (seconds) | Randomized Interval (seconds) |
+    |-----------|----------------------|----------------------------|
+    | 1         | 0.5                  | [0.25, 0.75]              |
+    | 2         | 0.75                 | [0.375, 1.125]            |
+    | 3         | 1.125                | [0.562, 1.687]            |
+    | 4         | 1.687                | [0.8435, 2.53]            |
+    | 5         | 2.53                 | [1.265, 3.795]            |
+    | 6         | 3.795                | [1.897, 5.692]            |
+    | 7         | 5.692                | [2.846, 8.538]            |
+    | 8         | 8.538                | [4.269, 12.807]           |
+    | 9         | 12.807               | [6.403, 19.210]           |
+    | 10        | 19.210               | 0                         |
+
+    Note: The sequence stops at request #10 when `max_retries` or `max_time_elapsed` is exceeded
+    """
+
+    initial_interval: float = Field(0.5, title="Initial backoff interval in seconds", gt=0)
+    randomization_factor: float = Field(0.5, title="Factor to randomize backoff", ge=0, le=1)
+    multiplier: float = Field(1.5, title="Multiply interval by this factor each retry", gt=0)
```

Comment:
```suggestion
    multiplier: float = Field(1.5, title="Multiply interval by this factor each retry", gt=1)
```
I think shrinking intervals are not expected from exponential backoff, should we limit this with validation? Thanks for pydantic btw! ❤️